File: Source/Aladdin.ASN1/ASN1/IObjectFactory.cs
namespace Aladdin.ASN1
{
	///////////////////////////////////////////////////////////////////////////
	// Создание объекта
	///////////////////////////////////////////////////////////////////////////
	public interface IObjectFactory
	{
		// проверить допустимость типа и объекта
		bool IsValidTag(Tag tag); void Validate(IEncodable obj, bool encode);

        // раскодировать объект
        IEncodable Decode(IEncodable decodable); 
	}
	///////////////////////////////////////////////////////////////////////////
	// Создание объекта
	///////////////////////////////////////////////////////////////////////////
	public interface IObjectFactory<T> : IObjectFactory where T : IEncodable
	{
		// проверить корректность объекта
		void Validate(T obj, bool encode); 

        // раскодировать объект
        new T Decode(IEncodable decodable); 
	}
}

File: Source/Aladdin.ASN1.Java/src/aladdin/asn1/IObjectFactory.java
package aladdin.asn1;
import java.io.*; 

///////////////////////////////////////////////////////////////////////////
// Создание объекта
///////////////////////////////////////////////////////////////////////////
public interface IObjectFactory<T extends IEncodable>
{
    // проверить допустимость типа
    boolean isValidTag(Tag tag); 

    // создать объект
    T decode(IEncodable encodable) throws IOException;

    // проверить корректность объекта
    void validate(IEncodable obj, boolean encode) throws IOException;
}
